plugins/database/mysql/connection_producer_test.go

```diff
@@ -0,0 +1,272 @@
+	whoamiCmd := "SELECT CURRENT_USER()"
+
+	client, err := mysql.getConnection(ctx)
+	if err != nil {
+		t.Fatalf("Unable to make connection to MySQL: %s", err)
+	}
+	stmt, err := client.Prepare(whoamiCmd)
+	if err != nil {
+		t.Fatalf("Unable to prepare MySQL statementL %s", err)
+	}
+
+	results := stmt.QueryRow()
+
+	expected := fmt.Sprintf("%s@%s", username, "%")
+
+	var result string
+	results.Scan(&result)
+
+	if !reflect.DeepEqual(result, expected) {
+		t.Fatalf("Actual:%#v\nExpected:\n%#v", result, expected)
+	}
+}
+
+func makeTempDir(t *testing.T) (confDir string) {
+	confDir, err := ioutil.TempDir(".", "mysql-test-data")
+	if err != nil {
+		t.Fatalf("Unable to make temp directory: %s", err)
+	}
+	// Convert the directory to an absolute path because docker needs it when mounting
+	confDir, err = filepath.Abs(filepath.Clean(confDir))
+	if err != nil {
+		t.Fatalf("Unable to determine where temp directory is on absolute path: %s", err)
+	}
+	return confDir
+}
+
+func startMySQLWithTLS(t *testing.T, version string, confDir string) (retURL string, cleanup func()) {
+	if os.Getenv("MYSQL_URL") != "" {
+		return os.Getenv("MYSQL_URL"), func() {}
+	}
+
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		t.Fatalf("Failed to connect to docker: %s", err)
+	}
+	pool.MaxWait = 30 * time.Second
+
+	containerName := "mysql-unit-test"
+
+	// Remove previously running container if it is still running because cleanup failed
+	err = pool.RemoveContainerByName(containerName)
+	if err != nil {
+		t.Fatalf("Unable to remove old running containers: %s", err)
+	}
+
+	runOpts := &dockertest.RunOptions{
+		Name:       containerName,
+		Repository: "mysql",
+		Tag:        version,
+		Cmd:        []string{"--defaults-extra-file=/etc/mysql/my.cnf", "--auto-generate-certs=OFF"},
+		Env:				[]string{"MYSQL_ROOT_PASSWORD=x509test"},
+		// Mount the directory from local filesystem into the container
+		Mounts: []string{
+			fmt.Sprintf("%s:/etc/mysql", confDir),
+		},
+	}
+
+	resource, err := pool.RunWithOptions(runOpts)
+	if err != nil {
+		t.Fatalf("Could not start local mysql docker container: %s", err)
+	}
+	resource.Expire(600)
+
+	cleanup = func() {
+		err := pool.Purge(resource)
+		if err != nil {
+			t.Fatalf("Failed to cleanup local container: %s", err)
+		}
+	}
+
+	dsn := fmt.Sprintf("{{username}}:{{password}}@tcp(localhost:%s)/mysql", resource.GetPort("3306/tcp"))
+
+	url := dbutil.QueryHelper(dsn, map[string]string{
+		"username": "root",
+		"password": "x509test",
+	})
+	// exponential backoff-retry
+	err = pool.Retry(func() error {
+		var err error
+
+		db, err := sql.Open("mysql", url)
+		if err != nil {
+			t.Logf("err: %s", err)
+			return err
+		}
+		defer db.Close()
+		return db.Ping()
+	})
+	if err != nil {
+		cleanup()
+		t.Fatalf("Could not connect to mysql docker container: %s", err)
+	}
+
+	return dsn, cleanup
+}
+
+func connect(t *testing.T, dsn string) (db *sql.DB) {
+	url := dbutil.QueryHelper(dsn, map[string]string{
```

Comment:
It shouldn't be. The username and password shou`{{username}}:{{password}}` respectively because the same dsn is used for all connections to the DB because the port number is only obtained as part of the container builder.

Reply:
It looks like `{{username}}` and `{{password}}` are being replaced in `Init` though: https://github.com/hashicorp/vault/pull/9181/files/702e4ed83b0e8379eba0e76c1416dcbcf770de73#diff-6b456b96c2b47338d3927b58a487f0ecR75-R78